### DISCUSSION TITLE
Como sería el uso de Facade?

### DISCUSSION THREAD

**Discussion by maribelcalla** [JosephCalla/Design-Patterns-Swift#9] [Q&A] [answered]
Sería de gran ayuda el uso de Facade. Gracias de ante mano.

**Comment by JosephCalla** [comment, marked as answer]
Facade is a structural design pattern that provides a simplified interface to a library, a framework, or any other complex set of classes.

Problem

Imagine that you must make your code work with a broad set of objects that belong to a sophisticated library or framework. Ordinarily, you’d need to initialize all of those objects, keep track of dependencies, execute methods in the correct order, and so on.

As a result, the business logic of your classes would become tightly coupled to the implementation details of 3rd-party classes, making it hard to comprehend and maintain

Solution

A facade is a class that provides a simple interface to a complex subsystem which contains lots of moving parts. A facade might provide limited functionality in comparison to working with the subsystem directly. However, it includes only those features that clients really care about.

Having a facade is handy when you need to integrate your app with a sophisticated library that has dozens of features, but you just need a tiny bit of its functionality.

For instance, an app that uploads short funny videos with cats to social media could potentially use a professional video conversion library. However, all that it really needs is a class with the single method encode(filename, format). After creating such a class and connecting it with the video conversion library, you’ll have your first facade.

Screenshot 2022-11-28 at 05 52 33

```
protocol CreditModule {
    func showCredit()
}

class BlackModule: CreditModule {
    func showCredit() {
        print("La tarjeta Black tiene un credito de un 1.000.000")
    }
}

class GoldModule: CreditModule {
    func showCredit() {
        print("La tarjeta Gold tiene un credito de 50.000")
    }
}

class SilverModule: CreditModule {
    func showCredit() {
        print("La tarjeta Silver tiene un credito de 50.000")
    }
}

class CreditMarketFacade {
    var black = BlackModule()
    var gold = GoldModule()
    var silver = SilverModule()
    
    func showCreditBlack() {
        black.showCredit()
    }
    
    func showCreditGold() {
        gold.showCredit()
    }
    
    func showCreditSilver() {
        silver.showCredit()
    }
    
}

/// Test
func testFacade() {
    let facate = CreditMarketFacade()
    facate.showCreditBlack()
    facate.showCreditGold()
    facate.showCreditSilver()
}

testFacade()
```
